// att-dhs/js/att.dhs.js
/*jslint browser: true, devel: true, node: true, debug: true, todo: true, indent: 2, maxlen: 150, unparam: true*/
/*global exports, require*/

'use strict';

var config = require('./att.dhs.config.js').config,
  webSocket = require('./att.dhs.websocket').webSocket,
  routes = require('./att.dhs.routes.js').routes;

config.initialize({
  config_route: routes.getConfig,
  tokens_route: routes.createToken,
  e911ids_route: routes.createE911Id,
  create_socket: webSocket.createSocket
});

module.exports = {
  /**
   * @module
   * dhs
   *
   *@desc
   *
   * The `dhs` namespace provides a client API for using our optional DHS RESTful API.
   * The DHS RESTful API allows you to perform the following actions:
   *
   * * Create access tokens using AT&T's OAuth
   * * Create E911 Ids using AT&T's OAuth and API
   */
  configure: config.configure,
  getConfig: routes.getConfig,
  createToken: routes.createToken,
  createE911Id: routes.createE911Id,
  postWebSocket: routes.postWebSocket,
  use: config.use
};
